File: app/src/main/java/com/productivity/habits/ui/daily/NumericHabitControls.kt
package com.productivity.habits.ui.daily

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Edit
import androidx.compose.material.icons.filled.Remove
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.productivity.habits.data.local.entity.HabitEntity
import com.productivity.habits.domain.engine.DynamicStepEngine
import com.productivity.habits.ui.common.HapticsHelper
import java.text.NumberFormat
import java.util.Locale

@Composable
fun NumericHabitControls(
    habit: HabitEntity,
    currentValue: Double,
    isCompleted: Boolean,
    accentColor: Color,
    onValueChange: (Double) -> Unit,
    onDeltaAdd: (Double) -> Unit,
    modifier: Modifier = Modifier
) {
    val context = LocalContext.current
    val haptic = LocalHapticFeedback.current
    var showDirectInputDialog by remember { mutableStateOf(false) }

    val targetValue = habit.targetValue ?: 1.0
    val unit = habit.unit ?: ""
    val stepConfig = DynamicStepEngine.getDynamicStepConfig(targetValue, habit.unit)

    val progress = (currentValue / targetValue).coerceIn(0.0, 1.0).toFloat()
    val animatedProgress by animateFloatAsState(targetValue = progress, label = "NumericProgress")

    val numberFormatter = NumberFormat.getNumberInstance(Locale.getDefault()).apply {
        maximumFractionDigits = 1
    }

    if (showDirectInputDialog) {
        DirectNumericInputDialog(
            habitTitle = habit.title,
            currentValue = currentValue,
            targetValue = targetValue,
            unit = habit.unit,
            onDismiss = { showDirectInputDialog = false },
            onConfirm = { newValue ->
                showDirectInputDialog = false
                val previouslyMet = currentValue >= targetValue
                onValueChange(newValue)
                if (!previouslyMet && newValue >= targetValue) {
                    HapticsHelper.performHeavyConfirmationHaptic(context, haptic)
                } else {
                    HapticsHelper.performLightHaptic(haptic)
                }
            }
        )
    }

    Column(modifier = modifier.fillMaxWidth()) {
        // Progress Row: Label + Edit Pencil
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = numberFormatter.format(currentValue),
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold,
                    color = if (isCompleted) accentColor else MaterialTheme.colorScheme.onSurface
                )
                Text(
                    text = " / ${numberFormatter.format(targetValue)} $unit".trimEnd(),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.padding(start = 4.dp)
                )
            }

            IconButton(
                onClick = {
                    HapticsHelper.performLightHaptic(haptic)
                    showDirectInputDialog = true
                },
                modifier = Modifier.size(28.dp)
            ) {
                Icon(
                    imageVector = Icons.Default.Edit,
                    contentDescription = "Edit value",
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.size(16.dp)
                )
            }
        }

        Spacer(modifier = Modifier.height(6.dp))

        // Progress Bar
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(6.dp)
                .clip(RoundedCornerShape(3.dp))
                .background(MaterialTheme.colorScheme.surfaceVariant)
        ) {
            Box(
                modifier = Modifier
                    .fillMaxWidth(animatedProgress)
                    .fillMaxHeight()
                    .clip(RoundedCornerShape(3.dp))
                    .background(accentColor)
            )
        }

        Spacer(modifier = Modifier.height(10.dp))

        // Stepper & Quick Add Chips Row
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            // Minus / Plus Steppers
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(6.dp)
            ) {
                // Minus primary step
                Surface(
                    modifier = Modifier
                        .size(32.dp)
                        .clip(CircleShape)
                        .clickable(enabled = currentValue > 0) {
                            HapticsHelper.performLightHaptic(haptic)
                            onDeltaAdd(-stepConfig.primaryStep)
                        },
                    shape = CircleShape,
                    color = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.7f),
                    border = BorderStroke(1.dp, MaterialTheme.colorScheme.outlineVariant)
                ) {
                    Box(contentAlignment = Alignment.Center) {
                        Icon(
                            imageVector = Icons.Default.Remove,
                            contentDescription = "Decrease",
                            tint = if (currentValue > 0) MaterialTheme.colorScheme.onSurface else MaterialTheme.colorScheme.outline,
                            modifier = Modifier.size(16.dp)
                        )
                    }
                }

                // Plus primary step
                Surface(
                    modifier = Modifier
                        .size(32.dp)
                        .clip(CircleShape)
                        .clickable {
                            val newTotal = currentValue + stepConfig.primaryStep
                            val wasMet = currentValue >= targetValue
                            if (!wasMet && newTotal >= targetValue) {
                                HapticsHelper.performHeavyConfirmationHaptic(context, haptic)
                            } else {
                                HapticsHelper.performLightHaptic(haptic)
                            }
                            onDeltaAdd(stepConfig.primaryStep)
                        },
                    shape = CircleShape,
                    color = accentColor.copy(alpha = 0.15f),
                    border = BorderStroke(1.dp, accentColor.copy(alpha = 0.3f))
                ) {
                    Box(contentAlignment = Alignment.Center) {
                        Icon(
                            imageVector = Icons.Default.Add,
                            contentDescription = "Increase",
                            tint = accentColor,
                            modifier = Modifier.size(18.dp)
                        )
                    }
                }
            }

            // Quick Add Chips
            Row(
                horizontalArrangement = Arrangement.spacedBy(6.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                stepConfig.quickAddValues.forEach { quickVal ->
                    val quickValFormatted = if (quickVal % 1.0 == 0.0) "+${quickVal.toInt()}" else "+$quickVal"
                    Surface(
                        modifier = Modifier
                            .clip(RoundedCornerShape(14.dp))
                            .clickable {
                                val newTotal = currentValue + quickVal
                                val wasMet = currentValue >= targetValue
                                if (!wasMet && newTotal >= targetValue) {
                                    HapticsHelper.performHeavyConfirmationHaptic(context, haptic)
                                } else {
                                    HapticsHelper.performLightHaptic(haptic)
                                }
                                onDeltaAdd(quickVal)
                            },
                        shape = RoundedCornerShape(14.dp),
                        color = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.5f),
                        border = BorderStroke(1.dp, MaterialTheme.colorScheme.outlineVariant.copy(alpha = 0.6f))
                    ) {
                        Text(
                            text = quickValFormatted,
                            style = MaterialTheme.typography.labelSmall,
                            fontWeight = FontWeight.SemiBold,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                            modifier = Modifier.padding(horizontal = 8.dp, vertical = 4.dp)
                        )
                    }
                }
            }
        }
    }
}

File: lib/ui/daily/widgets/numeric_habit_controls.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import '../../../domain/engines/dynamic_step_engine.dart';
import '../../../domain/models/habit.dart';
import '../../common/haptics_helper.dart';
import '../dialogs/direct_numeric_input_dialog.dart';

class NumericHabitControls extends StatelessWidget {
  final Habit habit;
  final double currentValue;
  final bool isCompleted;
  final Color accentColor;
  final ValueChanged<double> onValueChange;
  final ValueChanged<double> onDeltaAdd;

  const NumericHabitControls({
    super.key,
    required this.habit,
    required this.currentValue,
    required this.isCompleted,
    required this.accentColor,
    required this.onValueChange,
    required this.onDeltaAdd,
  });

  void _showDirectInputDialog(BuildContext context, double targetValue) {
    showDialog(
      context: context,
      builder: (dialogCtx) => DirectNumericInputDialog(
        habitTitle: habit.title,
        currentValue: currentValue,
        targetValue: targetValue,
        unit: habit.unit,
        onDismiss: () => Navigator.of(dialogCtx).pop(),
        onConfirm: (newValue) {
          Navigator.of(dialogCtx).pop();
          final previouslyMet = currentValue >= targetValue;
          onValueChange(newValue);
          if (!previouslyMet && newValue >= targetValue) {
            HapticsHelper.performHeavyConfirmationHaptic();
          } else {
            HapticsHelper.performLightHaptic();
          }
        },
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final targetValue = habit.targetValue ?? 1.0;
    final unit = habit.unit ?? '';
    final stepConfig =
        DynamicStepEngine.getDynamicStepConfig(targetValue, habit.unit);

    final progress = (currentValue / targetValue).clamp(0.0, 1.0);

    final numberFormatter = NumberFormat.decimalPattern();

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        // Progress Row: Label + Edit Pencil
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Row(
              children: [
                Text(
                  numberFormatter.format(currentValue),
                  style: theme.textTheme.titleMedium?.copyWith(
                    fontWeight: FontWeight.bold,
                    color: isCompleted
                        ? accentColor
                        : theme.colorScheme.onSurface,
                  ),
                ),
                Padding(
                  padding: const EdgeInsets.only(left: 4),
                  child: Text(
                    ' / ${numberFormatter.format(targetValue)} $unit'.trimRight(),
                    style: theme.textTheme.bodySmall?.copyWith(
                      color: theme.colorScheme.onSurfaceVariant,
                    ),
                  ),
                ),
              ],
            ),
            IconButton(
              iconSize: 16,
              constraints: const BoxConstraints(
                minWidth: 28,
                minHeight: 28,
              ),
              padding: EdgeInsets.zero,
              icon: Icon(
                Icons.edit,
                color: theme.colorScheme.onSurfaceVariant,
              ),
              tooltip: 'Edit value',
              onPressed: () {
                HapticsHelper.performLightHaptic();
                _showDirectInputDialog(context, targetValue);
              },
            ),
          ],
        ),
        const SizedBox(height: 6),

        // Progress Bar
        ClipRRect(
          borderRadius: BorderRadius.circular(3),
          child: SizedBox(
            height: 6,
            child: LinearProgressIndicator(
              value: progress,
              backgroundColor: theme.colorScheme.surfaceContainerHighest,
              valueColor: AlwaysStoppedAnimation<Color>(accentColor),
            ),
          ),
        ),
        const SizedBox(height: 10),

        // Stepper & Quick Add Chips Row
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            // Minus / Plus Steppers
            Row(
              children: [
                // Minus primary step
                Material(
                  color: theme.colorScheme.surfaceContainerHighest
                      .withValues(alpha: 0.7),
                  shape: CircleBorder(
                    side: BorderSide(
                      color: theme.colorScheme.outlineVariant,
                    ),
                  ),
                  child: InkWell(
                    customBorder: const CircleBorder(),
                    onTap: currentValue > 0
                        ? () {
                            HapticsHelper.performLightHaptic();
                            onDeltaAdd(-stepConfig.primaryStep);
                          }
                        : null,
                    child: SizedBox(
                      width: 32,
                      height: 32,
                      child: Center(
                        child: Icon(
                          Icons.remove,
                          size: 16,
                          color: currentValue > 0
                              ? theme.colorScheme.onSurface
                              : theme.colorScheme.outline,
                        ),
                      ),
                    ),
                  ),
                ),
                const SizedBox(width: 6),

                // Plus primary step
                Material(
                  color: accentColor.withValues(alpha: 0.15),
                  shape: CircleBorder(
                    side: BorderSide(
                      color: accentColor.withValues(alpha: 0.3),
                    ),
                  ),
                  child: InkWell(
                    customBorder: const CircleBorder(),
                    onTap: () {
                      final newTotal = currentValue + stepConfig.primaryStep;
                      final wasMet = currentValue >= targetValue;
                      if (!wasMet && newTotal >= targetValue) {
                        HapticsHelper.performHeavyConfirmationHaptic();
                      } else {
                        HapticsHelper.performLightHaptic();
                      }
                      onDeltaAdd(stepConfig.primaryStep);
                    },
                    child: SizedBox(
                      width: 32,
                      height: 32,
                      child: Center(
                        child: Icon(
                          Icons.add,
                          size: 18,
                          color: accentColor,
                        ),
                      ),
                    ),
                  ),
                ),
              ],
            ),

            // Quick Add Chips
            Row(
              children: stepConfig.quickAddValues.map((quickVal) {
                final quickValFormatted = quickVal % 1.0 == 0.0
                    ? '+${quickVal.toInt()}'
                    : '+$quickVal';

                return Padding(
                  padding: const EdgeInsets.only(left: 6),
                  child: Material(
                    color: theme.colorScheme.surfaceContainerHighest
                        .withValues(alpha: 0.5),
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(14),
                      side: BorderSide(
                        color: theme.colorScheme.outlineVariant
                            .withValues(alpha: 0.6),
                      ),
                    ),
                    child: InkWell(
                      borderRadius: BorderRadius.circular(14),
                      onTap: () {
                        final newTotal = currentValue + quickVal;
                        final wasMet = currentValue >= targetValue;
                        if (!wasMet && newTotal >= targetValue) {
                          HapticsHelper.performHeavyConfirmationHaptic();
                        } else {
                          HapticsHelper.performLightHaptic();
                        }
                        onDeltaAdd(quickVal);
                      },
                      child: Padding(
                        padding: const EdgeInsets.symmetric(
                          horizontal: 8,
                          vertical: 4,
                        ),
                        child: Text(
                          quickValFormatted,
                          style: theme.textTheme.labelSmall?.copyWith(
                            fontWeight: FontWeight.w600,
                            color: theme.colorScheme.onSurfaceVariant,
                          ),
                        ),
                      ),
                    ),
                  ),
                );
              }).toList(),
            ),
          ],
        ),
      ],
    );
  }
}
